File: HOT100/7.链表/141.环形链表.cpp
/*
 * @lc app=leetcode.cn id=141 lang=cpp
 *
 * [141] 环形链表
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (1)
        {
            if (fast && fast->next)
            {
                fast = fast->next->next;
                slow = slow->next;
            }
            else return 0;
            if (slow == fast) break;
        }
        if (fast) return 1;
        else return 0;
    }
};
// @lc code=end

File: CodeTop/字节/141.环形链表.py
#
# @lc app=leetcode.cn id=141 lang=python3
#
# [141] 环形链表
#
from typing import Optional
class ListNode:
    def __init__(self, val = 0, next = None) -> None:
        self.val = val
        self.next = next
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = fast = head
        while fast:
            slow = slow.next
            if not fast.next: return False
            fast = fast.next.next
            if slow == fast: return True
        
# @lc code=end

